File: Retos/Reto #4 - PRIMO, FIBONACCI Y PAR [Media]/dart/reanthonyh.dart
import 'dart:math';

void main() {
  print(evaluateNumber(2));
  print(evaluateNumber(7));
  print(evaluateNumber(11));
  print(evaluateNumber(13));
  print(evaluateNumber(34));
}

extension IntExtention on int {
  bool get isPrime {
    if (this <= 1) {
      return false;
    }
    for (var i = 2; i < sqrt(this); i++) {
      if (this % i == 0) {
        return false;
      }
    }
    return true;
  }

  bool get isFibonacci {
    if (this == 0 || this == 1 || this == 2) {
      return true;
    }

    final fibonacciNumbers = [0, 1];
    for (var i = 2; i < this; i++) {
      final fiboNumber = fibonacciNumbers[i - 2] + fibonacciNumbers[i - 1];
      if (fiboNumber == this) {
        return true;
      }
      fibonacciNumbers.insert(i, fiboNumber);
    }
    return false;
  }

  bool get isEven {
    return this % 2 == 0;
  }
}

String evaluateNumber(int number) {
  if (!number.isNegative) {
    String isPrime = number.isPrime ? 'es primo' : 'no es primo';
    String isFibonacci = number.isFibonacci ? 'fibonacci ' : 'no es fibonacci';
    String isEven = number.isEven ? 'par' : 'impar';

    return "$number $isPrime, $isFibonacci y es $isEven";
  }
  return "$number es negativo";
}

File: Retos/Reto #4 - PRIMO, FIBONACCI Y PAR [Media]/swift/reanthonyh.swift
import Foundation

evaluateNumber(3)
evaluateNumber(7)
evaluateNumber(11)
evaluateNumber(13)
evaluateNumber(34)

func evaluateNumber(_ number: Int) {
  let isPrime = number.isPrime() ? "es primo" : "no es primo"
  let isFibonacci = number.isFibonacci() ? "es fibonacci" : "no es fibonacci"
  let isEven = number.isEven() ? "es par" : "es impar"
  print("\(number) \(isPrime), \(isFibonacci) y \(isEven)")
}

extension Int {
  func isPrime() -> Bool {
    if self <= 1 {
      return false
    }

    let limit = Int(Double(self).squareRoot())
    if limit > 2 {
      for index in 2..<limit {
        if self % index == 0 {
          return false
        }
      }
    }
    return true
  }

  func isFibonacci() -> Bool {
    if self == 0 || self == 1 || self == 2 {
      return true
    }
    var fibonacciList: [Int] = [0, 1]
    for i in 2..<15 {
      let number = fibonacciList[i - 2] + fibonacciList[i - 1]

      if self == number { return true }
      if self < number { return false }

      fibonacciList.append(number)
    }
    print("fibo: \(fibonacciList)")
    return false
  }

  func isEven() -> Bool {
    return self.isMultiple(of: 2)

  }
}
